### PR TITLE
New error system

### PR DESCRIPTION
**Pull request by ajfriend**
# notes

- This PR helped me spot and (I think) fix several memory allocation bugs. These bugs all followed the same pattern, and will be (or would have been) fixed by #261

### CODE REVIEW

**Comment by ajfriend** [comment]
After playing with this, I think it makes sense to land the changes I'm attempting in #261

File: src/h3/_cy/cells.pyx [hunk 1]
```diff
@@ -125,25 +129,22 @@ cpdef H3int[:] ring(H3int h, int k):
...
-    # the memory. context manager would be better here, if we can figure out
-    # how to do that
-    mv = create_mv(ptr, n)
-
-    if err:
+    try:
+        check_for_error(
+            h3lib.gridRingUnsafe(h, k, ptr)
+        )
+    except H3ValueError:
```
Comment (isaacbrodsky) [inline src/h3/_cy/cells.pyx:142]: I would expect try/except to be relatively slow here compared to looking at the error value itself. That may not matter for this PR.

File: src/h3/_cy/error_system.pyx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    pass
+
+cdef check_for_error(H3Error err):
+    """
+    todo: more descriptive message
+    todo: add error codes as property
+    todo: allow passing in extra args for the error message?
+    """
+
+    d = {
```
Comment (isaacbrodsky) [inline src/h3/_cy/error_system.pyx:36]: Does this require allocating the error objects and dictionary each time a check is made?

File: src/h3/_cy/h3lib.pxd [hunk 3]
```diff
@@ -2,9 +2,27 @@ from cpython cimport bool
 from libc.stdint cimport uint64_t, int64_t, uint32_t
 
 ctypedef uint64_t H3int
-ctypedef uint32_t H3Error
 ctypedef basestring H3str
 
+# todo: extern version of this?
+cdef enum H3Error:
```
Comment (isaacbrodsky) [inline src/h3/_cy/h3lib.pxd:8]: is there a way to inform the Cython compiler of the bit width of this enum?

File: src/h3/_cy/error_system.pyx (hunk 2 above)
Comment (ajfriend) [reply src/h3/_cy/error_system.pyx:36]: Maybe. But there's also a chance Cython optimizes that away. I know they do that with if-else chains by converting them to case statements when possible. We should definitely change this later for performance if that isn't the case tho. At this point, I feel like I'm still just prototyping the idea of what the error interface looks like.

File: src/h3/_cy/h3lib.pxd (hunk 3 above)
Comment (ajfriend) [reply src/h3/_cy/h3lib.pxd:8]: Not sure yet. The confusion between the error type and the error codes was what led me to https://github.com/uber/h3/pull/613

Not sure what the most elegant solution is here, but definitely open to ideas!

**Comment by ajfriend** [comment]
@isaacbrodsky, is it fair to think of the H3 error codes in a hierarchy? I'm now thinking the simplest thing is to have as close to a 1-1 mapping between the C error codes and the Python exceptions. We might also want to have some of the error classes inherit from the appropriate builtin Python exceptions, like `ValueError` and `MemoryError`.

Would this hierarchy make sense?

```
E_FAILED: H3Exception
    DNE: H3ValueError

        E_DOMAIN: H3DomainError
            E_LATLNG_DOMAIN: H3LatLngDomainError
            E_RES_DOMAIN: H3ResDomainError

        E_CELL_INVALID: H3CellInvalidError
        E_DIR_EDGE_INVALID: H3DirEdgeInvalidError
        E_UNDIR_EDGE_INVALID: H3UndirEdgeInvalidError
        E_VERTEX_INVALID: H3VertexInvalidError
        E_PENTAGON: H3PentagonError
        E_DUPLICATE_INPUT: H3DuplicateInputError
        E_NOT_NEIGHBORS: H3NotNeighborsError
        E_RES_MISMATCH: H3ResMismatchError
        E_OPTION_INVALID: H3OptionError

    DNE: H3MemoryError
        E_MEMORY: H3MemoryAllocationError
        E_MEMORY_BOUNDS: H3MemoryBoundsError
```
with something like
```
class H3Exception(Exception):
    pass

class H3ValueError(H3Exception, ValueError):
    pass

class H3MemoryError(H3Exception, MemoryError):
    pass
```

One weird thing is that we'd be creating two Python exceptions that don't map to a specific error code.

Q: Could we do something like rename `E_MEMORY` to `E_MEMORY_ALLOC`, and then use `E_MEMORY` as a parent of `E_MEMORY_ALLOC` and `E_MEMORY_BOUNDS`?

Q: Would it make sense to add an `E_VALUE`, that would basically be a parent error to all the non-memory-related H3 errors?

If we were to add those two, we could have a perfect 1-1 mapping between the C error codes and the H3-py exceptions.

**Comment by isaacbrodsky** [comment]
I'm not sure it makes sense to think of the errors as a hierarchy. I can see certain errors being "programmer errors" (E_MEMORY_BOUNDS, E_OPTION_INVALID, E_FAILED) vs "data errors" (E_LATLNG_DOMAIN, E_CELL_INVALID, etc) but I am not sure a more complete hierarchy makes sense. I think it would be unique to the Python bindings in any case.

**Comment by ajfriend** [comment]
> I'm not sure it makes sense to think of the errors as a hierarchy. I can see certain errors being "programmer errors" (E_MEMORY_BOUNDS, E_OPTION_INVALID, E_FAILED) vs "data errors" (E_LATLNG_DOMAIN, E_CELL_INVALID, etc) but I am not sure a more complete hierarchy makes sense.

Fair that this might be a Python thing. The Python error hierarchy is totally the thing that has me thinking about it.

Here's my hierarchy argument: The current error language already implies *some* hierarchy.

- `E_FAILED` is the parent of all other H3 errors. 
- `E_DOMAIN` is explained with: "Argument was outside of acceptable range (**when a more specific error code is not available**)"

So I don't think there's a debate around wether a hierarchy exists. It already does.

In this instance, it's just a question of wether it makes sense to branch off memory errors and value errors. That question could be answered differently across Python and C, but I think it'd be cleaner if they were the same.

File: src/h3/_cy/cells.pyx (hunk 1 above)
Comment (ajfriend) [reply src/h3/_cy/cells.pyx]: I thought about it, and I think I've got a better approach with these latest changes. Let me know what you think!

File: src/h3/_cy/error_system.pyx (hunk 2 above)
Comment (ajfriend) [reply src/h3/_cy/error_system.pyx]: Also improved on this by moving the mapping outside of the function. LMKWYT!

**Comment by ajfriend** [comment]
With all the changes above, this is basically a fresh PR, ready for some initial reviews :)

**Comment by ajfriend** [comment]
Note, I did mess with some of the validation aggressiveness in this PR, but I should probably roll that back and defer any such changes to the discussion in https://github.com/uber/h3-py/issues/266

**Comment by ajfriend** [comment]
# some todo items i might revisit in a second pass

- [x] tests for `h3_error_code`
- [ ] tests for unknown error codes
- [ ] think through `__init__.py`
- [ ] think through `.pxd`
- [ ] document error system file
- [ ] terminology for `H3GridNavigationError`?

**Comment by ajfriend** [comment]
I think this is ready for review. I isn't perfect, but I think it counts as forward progress. I'm thinking I'd like to get this landed (if it looks like a reasonable start to everyone), and then build off of these changes in #261.

After landing #261, I'll come back and revisit the design here.

File: src/h3/_cy/cells.pyx [hunk 4]
```diff
@@ -305,53 +301,51 @@ cpdef double mean_hex_area(int resolution, unit='km^2') except -1:
...
-        h3lib.H3Error err
         int64_t n
 
-    check_cell(start)
-    check_cell(end)
-
+    # todo: can we segfault here with invalid inputs?
+    # Can we trust the c library to validate the start/end cells?
+    # probably applies to all size/work pairs of functions...
     err = h3lib.gridPathCellsSize(start, end, &n)
```
Comment (isaacbrodsky) [inline src/h3/_cy/cells.pyx:340]: I'm not sure what this comments means

File: src/h3/_cy/cells.pyx [hunk 5]
```diff
@@ -305,53 +301,51 @@ cpdef double mean_hex_area(int resolution, unit='km^2') except -1:
...
     else:
-        raise H3ValueError('Unknown unit: {}'.format(unit))
+        raise ValueError('Unknown unit: {}'.format(unit))
+
+    check_for_error(err)
 
     return area
 
 
+cdef couldnt_find_line(err, start, end):
```
Comment (isaacbrodsky) [inline src/h3/_cy/cells.pyx:327]: It might be best to avoid contractions in function names

File: src/h3/_cy/error_system.pyx [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    pass
+
+
+@contextmanager
+def the_error(obj):
+    """
+    Syntactic maple syrup for grouping exception definitions.
+    The associated `with` statement ends up as a not-half-bad
+    approximation to a valid sentence fragment.
```
Comment (isaacbrodsky) [inline src/h3/_cy/error_system.pyx:113]: Nit as this does not effect anything about correctness, but this seems a little too cute for my taste. At a minimum I would add an underscore prefix to discourage importing this elsewhere.

File: src/h3/_cy/error_system.pyx [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+# TODO: Move the helpers to util?
+# TODO: Unclear how/where to expose these functions. cdef/cpdef?
+
+cdef code_to_exception(H3Error err):
+    if err == E_SUCCESS:
+        return None
+    elif err in error_mapping:
+        return error_mapping[err]
+    else:
+        raise UnknownH3ErrorCode(err)
```
Comment (isaacbrodsky) [inline src/h3/_cy/error_system.pyx:220]: It seems surprising that UnknownH3ErrorCode does not seem to derive from H3BaseException

File: src/h3/api/_api_template.py [hunk 8]
```diff
@@ -254,7 +254,7 @@ def h3_distance(self, h1, h2):
         path between the cells in the graph formed by connecting
         adjacent cells.
 
-        This function will return an H3ValueError if the
+        This function will return an exception if the
```
Comment (isaacbrodsky) [inline src/h3/api/_api_template.py:257]: I assume this should read "raise an exception"?

File: src/h3/_cy/error_system.pyx (hunk 7 above)
Comment (ajfriend) [reply src/h3/_cy/error_system.pyx:220]: Makes sense. I'll change it.

File: src/h3/_cy/cells.pyx (hunk 5 above)
Comment (ajfriend) [reply src/h3/_cy/cells.pyx:327]: Agreed.

File: src/h3/_cy/error_system.pyx (hunk 6 above)
Comment (ajfriend) [reply src/h3/_cy/error_system.pyx]: Definitely guilty of being cute. :)

I'm fine with whatever name suggestions you might have. I mostly just like the code construct for how clear and short
it makes the class hierarchy definitions:

```python
```

File: src/h3/_cy/cells.pyx (hunk 4 above)
Comment (ajfriend) [reply src/h3/_cy/cells.pyx:340]: Is it possible to segfault with a pair of calls to `gridPathCellsSize`/`gridPathCells`? If this function were fuzzed, for example?

If we can't segfault and just get junk output, then for the Python bindings we have a choice to validate or not validate the inputs for the user.

However, if it is possible to segfault, we have no choice and we *must* validate the inputs well enough to guarantee that a segfault can't happen.

**Comment by ajfriend** [comment]
The more I work on this PR, the less happy I am about mixing Python exceptions and Cython. I'll revisit the design (https://github.com/uber/h3-py/issues/267), but I'll land for now, just to move this along and unblock the improvements in #261 

@isaacbrodsky, let me know if you have any other concerns, and I'll can make an issue to make sure it gets addressed when I revisit.

**Pull request merged**